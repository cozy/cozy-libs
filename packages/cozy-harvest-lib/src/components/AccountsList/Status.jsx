import React from 'react'
import PropTypes from 'prop-types'

import Icon from 'cozy-ui/transpiled/react/Icon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import palette from 'cozy-ui/transpiled/react/palette'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import { getErrorLocale } from '../../helpers/konnectors'
import FlowProvider from '../FlowProvider'

/**
 *
 * Display trigger's status.
 * Get error / running from FlowProvider (aka realtime)
 */
const Status = ({ t, trigger, konnector }) => {
  return (
    <FlowProvider initialTrigger={trigger}>
      {({ flow }) => {
        const { error, running } = flow.getState()
        const errorTitle = getErrorLocale(error, konnector, t, 'title')
        if (running) {
          return (
            <div className="u-flex u-flex-justify-center u-flex-items-center">
              <Spinner className="u-flex-shrink-0" />
            </div>
          )
        }
        if (error) {
          return (
            <div className="u-flex u-flex-justify-center u-flex-items-center">
              <span className="u-mr-half u-caption u-pomegranate">
                {errorTitle}
              </span>
              <Icon icon="warning" size={16} className="u-flex-shrink-0" />
            </div>
          )
        }
        return (
          <Icon
            icon="check"
            color={palette['malachite']}
            size={16}
            className="u-flex-shrink-0"
          />
        )
      }}
    </FlowProvider>
  )
}

Status.propTypes = {
  konnector: PropTypes.shape({
    name: PropTypes.string,
    vendor_link: PropTypes.string
  }).isRequired,
  trigger: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(Status)
