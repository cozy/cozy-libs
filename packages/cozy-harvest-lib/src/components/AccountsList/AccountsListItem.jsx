import React from 'react'
import PropTypes from 'prop-types'

import { Account } from 'cozy-doctypes'
import Card from 'cozy-ui/transpiled/react/Card'
import MidEllipsis from 'cozy-ui/transpiled/react/MidEllipsis'
import { Caption } from 'cozy-ui/transpiled/react/Text'

import Status from './Status'

export class AccountsListItem extends React.PureComponent {
  render() {
    const { account, onClick, konnector, trigger } = this.props
    const accountName = Account.getAccountName(account)
    const accountLogin = Account.getAccountLogin(account)
    /**
     * When the account is from an OAuth, we cat not have an accountLogin.
     * So have to check if accountLogin is not undefined before doing the check
     *  */
    const shouldShowAccountLogin = accountLogin && accountName !== accountLogin
    return (
      <Card
        className="u-flex u-flex-justify-between u-flex-items-center u-c-pointer u-maw-100"
        onClick={onClick}
      >
        <div className="u-flex-grow-1 u-flex-shrink-1 u-ov-hidden u-mr-1">
          <MidEllipsis text={accountName} />
          {shouldShowAccountLogin && (
            <Caption>
              <MidEllipsis text={accountLogin} />
            </Caption>
          )}
        </div>
        <Status konnector={konnector} trigger={trigger} />
      </Card>
    )
  }
}

AccountsListItem.propTypes = {
  account: PropTypes.object.isRequired,
  konnector: PropTypes.shape({
    name: PropTypes.string,
    vendor_link: PropTypes.string
  }).isRequired,
  trigger: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired
}

export default AccountsListItem
