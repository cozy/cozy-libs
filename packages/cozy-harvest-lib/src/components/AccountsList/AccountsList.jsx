import React from 'react'
import PropTypes from 'prop-types'

import Card from 'cozy-ui/transpiled/react/Card'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import AccountsListItem from './AccountsListItem'

export class AccountsList extends React.PureComponent {
  render() {
    const { accounts, konnector, addAccount, onPick, t } = this.props

    return (
      <ul className="u-nolist u-p-0 u-m-0">
        {accounts.map((account, index) => (
          <li key={index} className="u-mb-half">
            <AccountsListItem
              account={account.account}
              konnector={konnector}
              trigger={account.trigger}
              onClick={() => onPick(account)}
            />
          </li>
        ))}
        <li className="u-mb-half">
          <Card className="u-p-0">
            <Button
              label={t('modal.addAccount.button')}
              theme="text"
              extension="full"
              icon="plus"
              className="u-bdrs-4"
              onClick={addAccount}
            />
          </Card>
        </li>
      </ul>
    )
  }
}

AccountsList.propTypes = {
  accounts: PropTypes.array.isRequired,
  konnector: PropTypes.shape({
    name: PropTypes.string,
    vendor_link: PropTypes.string
  }).isRequired,
  addAccount: PropTypes.func.isRequired,
  onPick: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(AccountsList)
