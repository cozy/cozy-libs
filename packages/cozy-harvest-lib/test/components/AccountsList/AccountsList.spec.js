import React from 'react'
import { shallow } from 'enzyme'

import { AccountsList } from 'components/AccountsList/AccountsList'

describe('AccountsList', () => {
  it('should render', () => {
    const wrapper = shallow(
      <AccountsList
        accounts={[
          {
            account: { _id: 'account-1' },
            trigger: { _id: 'trigger-1' }
          },
          {
            account: { _id: 'account-2' },
            trigger: { _id: 'trigger-2' }
          }
        ]}
        konnector={{
          name: 'test-konnector',
          vendor_link: 'test konnector link'
        }}
        addAccount={jest.fn()}
        onPick={jest.fn()}
        t={jest.fn(str => str)}
      />
    )
    const component = wrapper.shallow()
    expect(component.getElement()).toMatchSnapshot()
  })
})
