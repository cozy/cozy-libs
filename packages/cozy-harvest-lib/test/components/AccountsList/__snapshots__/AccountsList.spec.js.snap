// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsList should render 1`] = `
<ul
  className="u-nolist u-p-0 u-m-0"
>
  <li
    className="u-mb-half"
  >
    <AccountsListItem
      account={
        Object {
          "_id": "account-1",
        }
      }
      konnector={
        Object {
          "name": "test-konnector",
          "vendor_link": "test konnector link",
        }
      }
      onClick={[Function]}
      trigger={
        Object {
          "_id": "trigger-1",
        }
      }
    />
  </li>
  <li
    className="u-mb-half"
  >
    <AccountsListItem
      account={
        Object {
          "_id": "account-2",
        }
      }
      konnector={
        Object {
          "name": "test-konnector",
          "vendor_link": "test konnector link",
        }
      }
      onClick={[Function]}
      trigger={
        Object {
          "_id": "trigger-2",
        }
      }
    />
  </li>
  <li
    className="u-mb-half"
  >
    <Card
      className="u-p-0"
      tag="div"
    >
      <DefaultButton
        className="u-bdrs-4"
        extension="full"
        icon="plus"
        label="modal.addAccount.button"
        onClick={[MockFunction]}
        theme="text"
      />
    </Card>
  </li>
</ul>
`;
